### PR TITLE
Update

### PR DESCRIPTION
**Pull request by hasimpson-usgs**

### CODE REVIEW

**Comment by jmfee-usgs** [comment]
- split output-edge arguments, with defaults for ports and tag:
  output-edge HOST
  output-edge-port WRITE_PORT
  output-edge-cwb-port WRITE_PORT
  output-edge-read-port READ_PORT
  output-edge-tag TAG
- split input_edge arguments, with defaults for port
  input-edge HOST
  input-edge-port READ_PORT
- change update argument default to False (so you don't need to manually set)
- remove update-realtime argument, always try to fill older gaps during update
- update starttime, endtime arguments to specify UTCDateTime as type; argparse will convert when arguments included.
- change Controller factory constructor to accept only inputFactory, outputFactory, algorithm arguments.  Rest are options to run or run_update, geomag.py determines whether to call run or run_update.
- change Controller run and run_as_update method to accept one argument "options", pass args from geomag.py.

- update put_timeseries to use outchannels argument, instead of algorithm.get_output_channels.  optionally, verify algorithm.get_output_channels contains each channel in outchannels.

**Comment by hasimpson-usgs** [comment]
Can't do this.  It automatically instantiates UTCDateTime,  even when a
starttime/endtime pair isn't provided.  Which means that starttime/endtime
defaults to the current datetime.

   - update starttime, endtime arguments to specify UTCDateTime as type;
   argparse will convert when arguments included.




On Thu, Jul 2, 2015 at 11:35 AM, Jeremy Fee <notifications@github.com>
wrote:

>
>    - split output-edge arguments, with defaults for ports and tag:
>    output-edge HOST output-edge-port WRITE_PORT output-edge-cwb-port
>    WRITE_PORT output-edge-read-port READ_PORT output-edge-tag TAG
>    - split input_edge arguments, with defaults for port input-edge HOST
>    input-edge-port READ_PORT
>    - change update argument default to False (so you don't need to
>    manually set)
>    - remove update-realtime argument, always try to fill older gaps
>    during update
>    - update starttime, endtime arguments to specify UTCDateTime as type;
>    argparse will convert when arguments included.
>    - change Controller factory constructor to accept only inputFactory,
>    outputFactory, algorithm arguments. Rest are options to run or run_update,
>    geomag.py determines whether to call run or run_update.
>    -
>
>    change Controller run and run_as_update method to accept one argument
>    "options", pass args from geomag.py.
>    -
>
>    update put_timeseries to use outchannels argument, instead of
>    algorithm.get_output_channels. optionally, verify
>    algorithm.get_output_channels contains each channel in outchannels.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/usgs/geomag-algorithms/pull/32#issuecomment-118103607>
> .
>



-- 
Hal Simpson
IT Specialist
USGS Geomagnetism Program
Office:303-273-8459
Email:hasimpson@usgs.gov

**Comment by jmfee-usgs** [comment]
The argparse docs suggest it will only convert when the default is a string:
https://docs.python.org/2/library/argparse.html#default

**Comment by hasimpson-usgs** [comment]
Ready for review.   All changes made.  Seems to work.

On Mon, Jul 6, 2015 at 1:29 PM, Jeremy Fee <notifications@github.com> wrote:

> The argparse docs suggest it will only convert when the default is a
> string:
> https://docs.python.org/2/library/argparse.html#default
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/usgs/geomag-algorithms/pull/32#issuecomment-118970708>
> .
>



-- 
Hal Simpson
IT Specialist
USGS Geomagnetism Program
Office:303-273-8459
Email:hasimpson@usgs.gov

File: bin/geomag.py [hunk 1]
```diff
@@ -149,7 +153,19 @@ def main():
 
     controller = Controller(inputfactory, outputfactory, algorithm)
 
-    controller.run(UTCDateTime(args.starttime), UTCDateTime(args.endtime))
+    if args.starttime is not None:
```
Comment (jmfee-usgs) [inline bin/geomag.py]: - This shouldn't be needed anymore, now that argparse has type UTCDateTime.
- make other arguments, like `port` and `write_port` specify a type in argparse.

File: bin/geomag.py [hunk 2]
```diff
@@ -149,7 +153,19 @@ def main():
...
+        starttime = UTCDateTime(args.starttime)
+    else:
+        starttime = None
+    if args.endtime is not None:
+        endtime = UTCDateTime(args.endtime)
+    else:
+        endtime = None
+
+    if args.update:
+        controller.run_as_update(starttime, endtime, args)
```
Comment (jmfee-usgs) [inline bin/geomag.py]: only pass args into run, run_as_update, starttime/endtime are part of args.

File: geomagio/TimeseriesUtilities.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+    Returns
+    -------
+    array of gaps
+    """
+    gaps = []
+    gap = None
+
+    i = 0
+    data = stream[0].data
```
Comment (jmfee-usgs) [inline geomagio/TimeseriesUtilities.py]: should this handle multiple timeseries for the same channel, or have those been merged?

Comment (hasimpson-usgs) [inline geomagio/TimeseriesUtilities.py]: These are currently using the knowledge, that all data coming from the
input factory are merged,  and gaps are represented as Nan's.  Could add a
test to make certain that's true.

On Tue, Jul 7, 2015 at 9:33 AM, Jeremy Fee <notifications@github.com> wrote:

> In geomagio/TimeseriesUtilities.py
> <https://github.com/usgs/geomag-algorithms/pull/32#discussion_r34053719>:
>
> > +    stream: obspy.core.stream
> > +        a stream containing a single channel of data.
> > +    starttime: obspy.core.UTCDateTime
> > +        time of first sample.
> > +    endtime: obspy.core.UTCDateTime
> > +        time of last sample.
> > +
> > +    Returns
> > +    -------
> > +    array of gaps
> > +    """
> > +    gaps = []
> > +    gap = None
> > +
> > +    i = 0
> > +    data = stream[0].data
>
> should this handle multiple timeseries for the same channel, or have those
> been merged?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/usgs/geomag-algorithms/pull/32/files#r34053719>.
>



-- 
Hal Simpson
IT Specialist
USGS Geomagnetism Program
Office:303-273-8459
Email:hasimpson@usgs.gov

**Comment by jmfee-usgs** [comment]
Make the edge arguments include edge in their name, we may add support for other servers in the future, i.e. `edge_host`, `edge_port`, `edge_write_port`, etc.

**Comment by jmfee-usgs** [comment]
What is the difference between edge port and read_port?

**Comment by hasimpson-usgs** [comment]
In the off chance,  that somewhere in the future,  we are using update
between 2 different edges,  we would need to specify different ports.

On Tue, Jul 7, 2015 at 4:33 PM, Jeremy Fee <notifications@github.com> wrote:

> What is the difference between edge port and read_port?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/usgs/geomag-algorithms/pull/32#issuecomment-119363773>
> .
>



-- 
Hal Simpson
IT Specialist
USGS Geomagnetism Program
Office:303-273-8459
Email:hasimpson@usgs.gov

**Comment by jmfee-usgs** [comment]
I did a little refactoring in this branch, take a look and let me know what you think.  There is still a TODO to determine whether output_gaps can be processed when there are input_gaps.

https://github.com/jmfee-usgs/geomag-algorithms/tree/hasimpson-usgs-update

**Comment by hasimpson-usgs** [comment]
So,  2 cases I still see.

1) should we bother running a gap,  if there's no new data to fill it.
2) If we're updating data from another source (reading in an IAGA2002 file
from KAK) and we've determined there's a "block" gap to be merged,  we then
need to write only to the missing information in each channel.
Algorithms,  should never have this problem.

We can ignore both problems for now.  And return to them next week,  after
DeltaF is up and running.

On Tue, Jul 7, 2015 at 6:36 PM, Jeremy Fee <notifications@github.com> wrote:

> I did a little refactoring in this branch, take a look and let me know
> what you think. There is still a TODO to determine whether output_gaps can
> be processed when there are input_gaps.
>
> https://github.com/jmfee-usgs/geomag-algorithms/tree/hasimpson-usgs-update
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/usgs/geomag-algorithms/pull/32#issuecomment-119382584>
> .
>



-- 
Hal Simpson
IT Specialist
USGS Geomagnetism Program
Office:303-273-8459
Email:hasimpson@usgs.gov

**Comment by jmfee-usgs** [comment]
replaced by #36

**Pull request closed without merging**